Added 10 new userprofile fields to RosterEntry

pomegranited opened the pull request:
Adds the following fields from the to the learner `RosterEntry` model, and its corresponding `LearnerSerializer`.

* user_id
* language
* location
* year_of_birth
* level_of_education
* gender
* mailing_address
* city
* country
* goals

**Discussions**: This change is related to [edx-analytics-data-api PR #128](https://github.com/edx/edx-analytics-data-api/pull/128), which has been discussed extensively with the Analytics team.

**Dependencies**: The data for this change depends on [edx-analytics-pipeline PR #272](https://github.com/edx/edx-analytics-pipeline/pull/272).  This change can be installed without updating the analytics pipeline; however the new fields won't be populated until the updated pipeline tasks .

**Sandbox URL**: TBD - sandbox is being provisioned.

**Partner information**: 3rd party-hosted open edX instance

**Testing instructions**:

1. See [edx-analytics-data-api PR #128](https://github.com/edx/edx-analytics-data-api/pull/128) for testing setup instructions.  
1. If you want to see the data flow through the pipeline to the Analytics API, follow the testing instructions from [edx-analytics-pipeline PR #272](https://github.com/edx/edx-analytics-pipeline/pull/272).  Without this step, the new fields will display in the Analytics API, but they will be empty.
1. View the new learner data fields via the Analytics API:
    1. Using the swagger.io UI:
        1. Visit [the swagger.io UI](http://localhost:8100)
        1. Enter your token in the "API Key" field, and click Explore
        1. Try out GET /authenticated request to ensure 200 response returned
        1. Try out GET /api/v0/learners/ request, with your `course_id` filled in.
    1. Using `curl`:

      ```
API_TOKEN="<your API token>"
COURSE_ID="<your URL-encoded course_id>"

curl -X GET --header "Authorization: Token $API_TOKEN" "http://localhost:8100/api/v0/learners/?course_id=$COURSE_ID"
```

**Reviewers**
- [x] @haikuginger
- [ ] @mulby 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-analytics-data-api&number=129

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-1365](https://openedx.atlassian.net/browse/OSPR-1365) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-analytics-data-api/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

haikuginger commented:
:+1:

- [x] I tested this: I verified that the Analytics API reflected the values for the new fields provided by the Analytics pipeline.
- [x] I read through the code
- [x] Documentation was properly updated

pomegranited commented:
@mulby This issue is ready for review, whenever you get a chance.

dsjen commented:
@pomegranited -- thanks for the PR!  I'm going to start reviewing this shortly.

dsjen commented on analytics_data_api/v0/models.py:
```diff
@@ -249,9 +249,19 @@ def get_last_updated(cls):
 class RosterEntry(DocType):
 
     course_id = String()
+    user_id = Integer()
```
I think I asked this question in a google doc somewhere, but have forgotten the answer.  What is `user_id`?  How it is different from `username`?

pomegranited replied:
@dsjen The `user_id` is a numeric unique ID, used in the mysql databases as the user primary key.  We added it because our client requested it, but edX conventions (to which I can no longer find the link!) recommend using `username` as the identifier instead of `user_id`.

dsjen commented:
@pomegranited -- I tested this out and it looks good!  Just for completion, I did test this out where the new fields aren't explicitly set and the return results for string fields are not `null`/`None` but rather an empty string (e.g. `''`).  I think this is ok, and I just wanted to call it out in case it's unexpected behavior.

I do have one question about the purpose of the `user_id` field.  Otherwise, 👍 .  Thank you!

dsjen commented:
@pomegranited -- I just realized that the [management command to create the learners indices](https://github.com/edx/edx-analytics-data-api/blob/master/analytics_data_api/management/commands/create_elasticsearch_learners_indices.py) should be updated with the new fields added in this PR.  Would you please add those too?  Thanks!

dsjen commented:
Sorry about the merge conflicts here!  We upgraded to DRF 3.4.x in https://github.com/edx/edx-analytics-data-api/pull/135, and I believe some fields will need updating--shouldn't be too bad I think.

pomegranited commented:
@dsjen I've rebased, and added the new fields to the [management command to create the learners indices](https://github.com/edx/edx-analytics-data-api/pull/129/commits/4abcb7426792fd1ec8d97d80a30482c6e8f509fe), but I'm having trouble running it on the rebased code.

Can you run it and make sure it's ok?

And is there anything else I need to fix for this PR, or any other people who need to approve it?

Thank you!

dsjen commented:
@pomegranited -- thanks for adding the fields to the command and rebasing!  I tested it out and it looks great.  I'll merge it in shortly.

dsjen commented:
Oops, I forgot to ask that this be squashed before I very excitedly merged it.  We'll leave it as is, but didn't want to cause confusion whether squashing is preferable.

pomegranited commented:
No worries, @dsjen !  Wasn't too many commits anyway 😄   Thanks for merging.
